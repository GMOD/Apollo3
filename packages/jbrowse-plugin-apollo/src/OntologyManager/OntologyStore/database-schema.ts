/** schema types used to strongly-type using the `idb` type system */
import { DBSchema } from 'idb'

import {
  Edge as OboGraphEdge,
  Node as OboGraphNode,
} from './obo-graph-json-schema'

/** schema types used to strongly-type using the `idb` type system */
export interface OntologyDB extends DBSchema {
  meta: {
    key: string
    value: {
      id: string
      objectType: 'graph' | 'database'
      data: unknown
    }
  }
  nodes: {
    key: string
    value: OboGraphNode
  }
  edges: {
    key: string
    value: OboGraphEdge
  }
}
