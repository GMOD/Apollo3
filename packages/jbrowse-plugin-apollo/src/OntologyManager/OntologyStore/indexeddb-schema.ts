/** schema types used to strongly-type using the `idb` type system */
import { DBSchema } from 'idb/with-async-ittr'

import {
  Edge as OboGraphEdge,
  Meta as OboGraphMeta,
  Node as OboGraphNode,
} from './obo-graph-json-schema'
import { SourceLocation } from '.'

/** metadata about this IndexedDB ontology database */
export interface Meta {
  /** original OntologyManager record this was loaded from */
  ontologyRecord: {
    name: string
    version: string
    sourceLocation: SourceLocation
  }
  /** graph metadata in OBO Graph metadata format */
  graphMeta?: OboGraphMeta
  timestamp: string
  /** IndexedDB schemaVersion for this ontology */
  schemaVersion: number
  /** time taken to load this data */
  timings: {
    /** milliseconds to fetch, parse, and load the ontology */
    overall: number
    /** optional milliseconds to fetch the ontology source file over the network */
    fetch?: number
    /** optional milliseconds to parse the ontology source file */
    parse?: number
    /** optional milliseconds to load the ontology into IndexedDB */
    load?: number
  }
}

// a Node that goes in the DB must have id and type, so make a new type that specifies this
export type OntologyDBNode = OboGraphNode & {
  id: string
  type: 'CLASS' | 'INDIVIDUAL' | 'PROPERTY'
}
export function isOntologyDBNode(node: OboGraphNode): node is OntologyDBNode {
  return typeof node.id === 'string'
}

// an Edge that goes in our DB must have sub, pred, and obj, so make a new type and guard that specifies this
export type OntologyDBEdge = OboGraphEdge & {
  sub: string
  pred: string
  obj: string
}
export function isOntologyDBEdge(edge: OboGraphEdge): edge is OntologyDBEdge {
  return (
    typeof edge.sub === 'string' &&
    typeof edge.pred === 'string' &&
    typeof edge.obj === 'string'
  )
}

export function isDeprecated(thing: OntologyDBNode | OntologyDBEdge) {
  return Boolean(thing.meta?.deprecated)
}

/** schema types used to strongly-type using the `idb` type system */
export interface OntologyDB extends DBSchema {
  meta: {
    key: string
    value: Meta
  }
  nodes: {
    key: string
    value: OntologyDBNode
    indexes: {
      'by-label': string
      'by-type': string
      'by-synonym': string
    }
  }
  edges: {
    key: number
    value: OntologyDBEdge
    indexes: {
      'by-subject': string
      'by-object': string
      'by-predicate': string
    }
  }
}
