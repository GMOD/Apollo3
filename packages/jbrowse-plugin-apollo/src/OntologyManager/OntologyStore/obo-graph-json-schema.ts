/** TS types for OBO Graph JSON (https://github.com/geneontology/obographs) */
// first generated with https://transform.tools/json-schema-to-typescript, then edited for conciseness
export interface GraphDocument {
  '@context'?: unknown
  meta?: Meta
  graphs?: Graph[]
  [k: string]: unknown
}

export interface Meta {
  definition?: MetaDefinitionPropertyValue
  comments?: string[]
  subsets?: string[]
  synonyms?: MetaSynonymPropertyValue[]
  xrefs?: MetaXrefPropertyValue[]
  basicPropertyValues?: MetaBasicPropertyValue[]
  version?: string
  deprecated?: boolean
  [k: string]: unknown
}
export interface MetaDefinitionPropertyValue {
  pred?: string
  val?: string
  xrefs?: string[]
  meta?: Meta
  [k: string]: unknown
}
export interface MetaSynonymPropertyValue {
  synonymType?: string
  pred?: string
  val?: string
  xrefs?: string[]
  meta?: Meta
  [k: string]: unknown
}
export interface MetaXrefPropertyValue {
  lbl?: string
  pred?: string
  val?: string
  xrefs?: string[]
  meta?: Meta
  [k: string]: unknown
}
export interface MetaBasicPropertyValue {
  pred?: string
  val?: string
  xrefs?: string[]
  meta?: Meta
  [k: string]: unknown
}

export interface Graph {
  id?: string
  lbl?: string
  meta?: Meta
  nodes?: Node[]
  edges?: Edge[]

  equivalentNodesSets?: AxiomEquivalentNodesSet[]
  logicalDefinitionAxioms?: LogicalDefinitionAxiom[]
  domainRangeAxioms?: DomainRangeAxiom[]
  propertyChainAxioms?: PropertyChainAxiom[]
  [k: string]: unknown
}
export interface Node {
  id?: string
  lbl?: string
  type?: 'CLASS' | 'INDIVIDUAL' | 'PROPERTY'
  meta?: Meta
  [k: string]: unknown
}
export interface Edge {
  sub?: string
  pred?: string
  obj?: string
  meta?: Meta
  [k: string]: unknown
}
export interface AxiomEquivalentNodesSet {
  representativeNodeId?: string
  nodeIds?: string[]
  meta?: Meta
  [k: string]: unknown
}
export interface LogicalDefinitionAxiom {
  definedClassId?: string
  genusIds?: string[]
  restrictions?: AxiomExistentialRestrictionExpression[]
  meta?: Meta
  [k: string]: unknown
}
export interface AxiomExistentialRestrictionExpression {
  propertyId?: string
  fillerId?: string
  [k: string]: unknown
}
export interface DomainRangeAxiom {
  predicateId?: string
  domainClassIds?: string[]
  rangeClassIds?: string[]
  allValuesFromEdges?: Edge[]
  meta?: Meta
  [k: string]: unknown
}
export interface PropertyChainAxiom {
  predicateId?: string
  chainPredicateIds?: string[]
  meta?: Meta
  [k: string]: unknown
}
