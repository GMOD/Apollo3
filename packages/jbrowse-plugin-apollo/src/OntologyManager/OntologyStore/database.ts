import { openLocation } from '@jbrowse/core/util/io'
import { IDBPDatabase, IDBPTransaction, openDB } from 'idb'

import { OntologyDB } from './database-schema'
import GraphDocument from './obo-graph-json-schema'
import OntologyStore from '.'

/** schema version we are currently on, used for the IndexedDB schema open call */
const schemaVersion = 1

export type Database = IDBPDatabase<OntologyDB>

/** open the IndexedDB and create the DB schema if necessary */
export async function openDatabase(dbName: string) {
  return openDB<OntologyDB>(dbName, schemaVersion, {
    upgrade(
      database: IDBPDatabase<OntologyDB>,
      oldVersion: number,
      newVersion: number | null,
      transaction: IDBPTransaction<
        OntologyDB,
        ArrayLike<'nodes' | 'edges'>,
        'versionchange'
      >,
      event: IDBVersionChangeEvent,
    ): void {
      if (schemaVersion !== 1) {
        throw new Error(
          'now that the schemaVersion is past 1, you need to write some upgrade logic here',
        )
      }
      if (!database.objectStoreNames.contains('meta')) {
        database.createObjectStore('meta', { keyPath: 'id' })
      }
      if (!database.objectStoreNames.contains('nodes')) {
        database.createObjectStore('nodes', { keyPath: 'id' })
      }
      if (!database.objectStoreNames.contains('edges')) {
        database.createObjectStore('edges', { autoIncrement: true })
      }
    },
  })
}

/** load a OBO Graph JSON file into a database */
export async function loadOboGraphJson(store: OntologyStore, db: Database) {
  const startTime = Date.now()

  // TODO: using file streaming along with an event-based json parser
  // instead of JSON.parse and .readFile could probably make this faster
  // and less memory intensive
  const oboGraph = JSON.parse(
    await openLocation(store.sourceLocation).readFile('utf8'),
  ) as GraphDocument

  const parseTime = Date.now()

  const [graph, ...additionalGraphs] = oboGraph.graphs || []
  if (!graph) {
    return
  }
  if (additionalGraphs.length) {
    throw new Error('multiple graphs not supported')
  }

  debugger

  const tx = db.transaction(['meta', 'nodes', 'edges'], 'readwrite')

  // NOTE: all these .add promises are ignored because the transaction is
  // monitoring them and the tx.done promise will reject if they fail

  // load nodes
  const nodeStore = tx.objectStore('nodes')
  for (const node of graph.nodes || []) {
    await nodeStore.add(node)
  }
  // load edges
  const edgeStore = tx.objectStore('edges')
  for (const edge of graph.edges || []) {
    await edgeStore.add(edge)
  }

  // load graph meta data
  if (graph.meta) {
    const metaStore = tx.objectStore('meta')
    // graph metadata
    await metaStore.add({
      id: graph.id || 'graph',
      objectType: 'graph',
      data: graph.meta,
    })
  }

  await tx.done

  // record some metadata about this load operation
  const tx2 = db.transaction('meta', 'readwrite')
  void tx2.objectStore('meta').add({
    id: 'load',
    objectType: 'database',
    data: {
      timestamp: String(new Date()),
      schemaVersion,
      timings: {
        fetchAndParse: parseTime - startTime,
        loading: Date.now() - parseTime,
      },
    },
  })

  await tx2.done
  return
}

export async function isDatabaseCompletelyLoaded(db: Database) {
  // since metadata is loaded last, we use it as a signal that all the other data
  // was loaded
  const tx = db.transaction('meta')
  const count = await tx.objectStore('meta').count()
  return count !== 0
}
