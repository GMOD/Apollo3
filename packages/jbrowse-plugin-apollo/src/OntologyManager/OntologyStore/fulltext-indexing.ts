import { genericEnglishStopwords } from './fulltext-stopwords'
import { OntologyDBNode } from './indexeddb-schema'

function getStringAt(obj: Record<string, unknown>, path: string[]) {
  let thing = obj
  for (const key of path) {
    thing = thing[key] as Record<string, unknown>
  }
  return String(thing)
}

export function stringToWords(text: string) {
  return text.split(/[\W_]+/).filter((w) => Boolean(w))
}

/** @returns a Set of all the words at the given JSON paths that are not stopwords */
export function getWords(node: OntologyDBNode, paths: string[][]) {
  const wordSet = new Set<string>()
  for (const path of paths) {
    const text = getStringAt(node, path)
    stringToWords(text).forEach((word) => {
      const lc = word.toLowerCase()
      if (!genericEnglishStopwords.has(lc)) {
        wordSet.add(lc)
      }
    })
  }
  return wordSet
}
