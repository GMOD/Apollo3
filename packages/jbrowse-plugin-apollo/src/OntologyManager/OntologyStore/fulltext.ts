/* eslint-disable import/no-named-as-default-member */
// jsonpath triggers this rule for some reason. import { query } from 'jsonpath' does not work

import { checkAbortSignal } from '@jbrowse/core/util'
import jsonpath from 'jsonpath'

import { stopwords } from './fulltext-stopwords'
import { OntologyDBNode } from './indexeddb-schema'
import { applyPrefixes } from './prefixes'
import OntologyStore, { Transaction } from '.'
import { TextIndexFieldDefinition } from '..'

/** special value of jsonPath that gets the IRI (that is, ID) of the node with the configured prefixes applied */
export const PREFIXED_ID_PATH = '$PREFIXED_ID'

/** small wrapper for jsonpath.query that intercepts requests for the special prefixed ID path */
function jsonPathQuery(
  node: OntologyDBNode,
  path: string,
  prefixes: Map<string, string>,
) {
  if (path === PREFIXED_ID_PATH) {
    return [applyPrefixes(node.id, prefixes)]
  }
  return jsonpath.query(node, path)
}

function wordsInString(str: string) {
  return str
    .toLowerCase()
    .split(/[^A-Za-z0-9:]+/)
    .filter((word) => word && !stopwords.has(word))
}

/**
 * recursively get the indexable words from an iterator
 * of any objects
 **/
export function* extractWords(
  strings: Iterable<string>,
): Generator<string, void, undefined> {
  for (const str of strings) {
    yield* wordsInString(str)
  }
}

export function* extractStrings(
  things: Iterable<unknown>,
): Generator<string, void, undefined> {
  for (const thing of things) {
    if (typeof thing === 'string') {
      yield thing
    } else if (typeof thing === 'object') {
      const members = jsonpath.query(thing, '$..*')
      yield* extractStrings(members)
    }
  }
}

/** @returns generator of tuples of [jsonpath, word] */
export function* getWords(
  node: OntologyDBNode,
  jsonPaths: Iterable<string>,
  prefixes: Map<string, string>,
): Generator<[string, string], void, undefined> {
  for (const path of jsonPaths) {
    const queryResult = jsonPathQuery(node, path, prefixes) as unknown[]
    if (queryResult.length) {
      for (const word of extractWords(extractStrings(queryResult))) {
        yield [path, word]
      }
    }
  }
}

export interface Match {
  term: OntologyDBNode
  field: TextIndexFieldDefinition
  str: string
  score: number
}

export function isMatch(thing: object): thing is Match {
  return (
    'term' in thing && 'field' in thing && 'str' in thing && 'score' in thing
  )
}

/**
 *
 **/
export async function textSearch(
  this: OntologyStore,
  text: string,
  tx?: Transaction<['nodes']>,
  signal?: AbortSignal,
) {
  const myTx = tx ?? (await this.db).transaction(['nodes'])

  checkAbortSignal(signal)

  const queryWords = Array.from(wordsInString(text))

  const queries: Promise<void>[] = []

  /**
   * Build a structure of which terms match which words.
   * This is a Map of term.id -\> Set\<query word number\>
   **/
  const initialMatches = new Map<string, [OntologyDBNode, Set<number>]>()

  // find startsWith and complete matches
  queries.push(
    ...queryWords.map(async (queryWord, queryWordIndex) => {
      checkAbortSignal(signal)
      const idx = myTx.objectStore('nodes').index('full-text-words')
      for await (const cursor of idx.iterate(
        IDBKeyRange.bound(queryWord, `${queryWord}\uffff`, false, false),
      )) {
        checkAbortSignal(signal)
        const term = cursor.value
        const termMatches = initialMatches.get(term.id) ?? [
          term,
          new Set<number>(),
        ]
        termMatches[1].add(queryWordIndex)
        initialMatches.set(term.id, termMatches)
      }
    }),
  )

  await Promise.all(queries)

  checkAbortSignal(signal)

  // now rank the term matches and add some detail
  const results: Match[] = []
  for (const [, [term, wordIndexes]] of initialMatches) {
    checkAbortSignal(signal)
    results.push(
      ...elaborateMatch(
        this.textIndexFields,
        term,
        wordIndexes,
        queryWords,
        this.prefixes,
      ),
    )
  }

  // sort the terms by score descending
  results.sort((a, b) => b.score - a.score)

  // truncate if necessary
  return results.slice(
    0,
    this.options.maxSearchResults ?? this.DEFAULT_MAX_SEARCH_RESULTS,
  )
}

export function elaborateMatch(
  textIndexPaths: TextIndexFieldDefinition[],
  term: OntologyDBNode,
  queryWordIndexes: Set<number>,
  queryWords: string[],
  prefixes: Map<string, string>,
): Match[] {
  const sortedWordIndexes = Array.from(queryWordIndexes).sort()
  const matchedQueryWords = sortedWordIndexes.map((i) => queryWords[i])
  const queryWordRegexps = matchedQueryWords.map((queryWord) => {
    const escaped = queryWord.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
    return RegExp(`\\b${escaped}`, 'gi')
  })
  // const needle = matchedQueryWords.join(' ')

  // ranking weights that can be tweaked if you know what you're doing
  const FIELD_PRIORITY_WEIGHT = 1
  const MATCH_WORDS_CLOSENESS_WEIGHT = 0.05
  const MATCH_ADJACENCY_BONUS = 1
  const MATCH_RIGHT_ORDER_BONUS = 1
  const MATCH_LENGTH_WEIGHT = 0.01
  const PCT_OF_STRING_WEIGHT = 0.05
  const WORD_BONUS = 100 // bonus for each of the words matched

  // inspect the node at each of the index paths, because we don't know which ones matched
  interface WordMatch {
    wordIndex: number
    position: number
  }
  let matches: (Match & { wordMatches: WordMatch[] })[] = []
  let maxScore = 0
  textIndexPaths.forEach((field, fieldIdx) => {
    const wordsMatched = new Set<number>()
    const fieldPriorityBonus = textIndexPaths.length - fieldIdx - 1
    const termStrings = Array.from(
      extractStrings(jsonPathQuery(term, field.jsonPath, prefixes)),
    )
    // find occurrences of each of the words in the strings
    for (const str of termStrings) {
      let score = 0
      const wordMatches: WordMatch[] = []
      queryWordRegexps.forEach((re, wordIndex) => {
        for (const match of str.matchAll(re)) {
          score += 1 + fieldPriorityBonus * FIELD_PRIORITY_WEIGHT
          wordsMatched.add(wordIndex)
          const position = match.index
          const queryWord = queryWords[wordIndex]
          if (position !== undefined) {
            score += queryWord.length * MATCH_LENGTH_WEIGHT
            score +=
              (queryWord.length / str.length) * 100 * PCT_OF_STRING_WEIGHT
            wordMatches.push({ wordIndex, position })
          }
        }
      })

      // apply the words-matched bonus
      score += wordsMatched.size * WORD_BONUS

      if (maxScore < score) {
        maxScore = score
      }
      // sort the word matches by position in the target string ascending
      wordMatches.sort((a, b) => a.position - b.position)
      if (wordMatches.length) {
        matches.push({ term, field, str, score, wordMatches })
      }
    }
  })

  // Keep only the highest-scored matches. Usually 1, but there
  // could be multiple if there is a tie for first place.
  matches = matches.filter((m) => m.score === maxScore)

  for (const match of matches) {
    const { wordMatches } = match
    // re-examine the word order and spacing to give bonuses for the
    // right order and close spacing
    for (let i = 0; i < wordMatches.length - 1; i++) {
      // bonus for pairs with adjacent word indexes and close spacing
      const m1 = wordMatches[i]
      const m2 = wordMatches[i + 1]
      const wdiff = m2.wordIndex - m1.wordIndex
      if (wdiff === 1 || wdiff === -1) {
        // they are adjacent, bonus
        match.score += MATCH_ADJACENCY_BONUS
        if (wdiff === 1) {
          // they are in the right order, bonus
          match.score += MATCH_RIGHT_ORDER_BONUS
        }
        // give additional bonus for how close they are
        const spacing =
          Math.abs(
            m2.position -
              (m1.position + matchedQueryWords[m1.wordIndex].length),
          ) - 1
        match.score -= spacing * MATCH_WORDS_CLOSENESS_WEIGHT
      }
    }
  }

  return matches
}
