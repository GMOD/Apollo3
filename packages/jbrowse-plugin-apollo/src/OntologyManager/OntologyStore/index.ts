import {
  BlobLocation,
  LocalPathLocation,
  UriLocation,
  isUriLocation,
} from '@jbrowse/core/util'
import { IDBPTransaction } from 'idb'

import { OntologyDB } from './indexeddb-schema'
import {
  isDatabaseCompletelyLoaded,
  loadOboGraphJson,
  openDatabase,
} from './indexeddb-storage'

export type SourceLocation = UriLocation | LocalPathLocation | BlobLocation

type SourceType = 'obo-graph-json' | 'obo' | 'owl'

/**
 * @deprecated use the one from jbrowse core when it is published
 **/
function isLocalPathLocation(location: unknown): location is LocalPathLocation {
  return (
    typeof location === 'object' && location !== null && 'localPath' in location
  )
}

/**
 * @deprecated use the one from jbrowse core when it is published
 */
function isBlobLocation(location: unknown): location is BlobLocation {
  return (
    typeof location === 'object' && location !== null && 'blobId' in location
  )
}

/** query interface for a specific ontology */
export default class OntologyStore {
  ontologyName: string
  ontologyVersion: string
  sourceLocation: SourceLocation
  db: ReturnType<OntologyStore['prepareDatabase']>

  constructor(name: string, version: string, source: SourceLocation) {
    this.ontologyName = name
    this.ontologyVersion = version
    this.sourceLocation = source
    this.db = this.prepareDatabase()
  }

  /**
   * check that the configuration of this ontology appears valid. Does not
   * try to do any fetches, however.
   */
  validate() {
    const errors = []

    // validate the source's file type
    const { sourceType } = this
    if (!sourceType) {
      errors.push(
        new Error(
          `unable to determine format of ontology source file ${JSON.stringify(
            this.sourceLocation,
          )}, file name must end with ".json", ".obo", or ".owl"`,
        ),
      )
    } else if (sourceType !== 'obo-graph-json') {
      errors.push(
        new Error(
          `ontology source file ${JSON.stringify(
            this.sourceLocation,
          )} has type ${sourceType}, which is not yet supported`,
        ),
      )
    }

    return errors
  }

  get sourceType(): SourceType | undefined {
    if (isUriLocation(this.sourceLocation)) {
      if (this.sourceLocation.uri.endsWith('.json')) {
        return 'obo-graph-json'
      }
    } else if (isLocalPathLocation(this.sourceLocation)) {
      if (this.sourceLocation.localPath.endsWith('.json')) {
        return 'obo-graph-json'
      }
    }
    return undefined
  }

  /** base name of the IndexedDB database for this ontology */
  get dbName() {
    return `Apollo Ontology "${this.ontologyName}" "${this.ontologyVersion}"`
  }

  async prepareDatabase() {
    const errors = this.validate()
    if (errors.length) {
      throw errors
    }

    const db = await openDatabase(this.dbName)

    // if database is already completely loaded, just return it
    if (await isDatabaseCompletelyLoaded(db)) {
      return db
    }

    const { sourceType } = this
    if (sourceType === 'obo-graph-json') {
      await loadOboGraphJson(this, db)
    } else {
      throw new Error(
        `ontology source file ${JSON.stringify(
          this.sourceLocation,
        )} has type ${sourceType}, which is not yet supported`,
      )
    }

    return db
  }

  async nodeStore(tx?: IDBPTransaction<OntologyDB, ['nodes']>) {
    return (tx || (await this.db).transaction('nodes')).objectStore('nodes')
  }

  async nodeCount(tx?: IDBPTransaction<OntologyDB, ['nodes']>) {
    const nodes = await this.nodeStore(tx)
    return nodes.count()
  }

  async getTermsWithLabelOrSynonym(
    termLabelOrSynonym: string,
    tx?: IDBPTransaction<OntologyDB, ['nodes']>,
  ) {
    const nodes = await this.nodeStore(tx)
    const labeled = nodes.index('by-label').getAll(termLabelOrSynonym)
    const synonymed = await nodes.index('by-synonym').getAll(termLabelOrSynonym)
    return (await labeled).concat(synonymed)
  }

  async getValidPartsOf(termId: string) {
    return
  }

  async getPropertyAndSubPropertiesByLabel(propertyLabel: string) {
    const tx = (await this.db).transaction(['nodes', 'edges'])
    const edges = tx.objectStore('edges')
    const nodes = tx.objectStore('nodes')
    const mainProperties = (
      await this.getTermsWithLabelOrSynonym(
        propertyLabel,
        tx as unknown as IDBPTransaction<OntologyDB, ['nodes']>,
      )
    ).filter((t) => t.type === 'PROPERTY')
    const allProperties = (
      await Promise.all(
        mainProperties.map(async (propertyNode) => {
          const subPropertyIds = (
            await edges.index('by-object').getAll(propertyNode.id)
          )
            .filter((p) => p.pred === 'subPropertyOf')
            .map((p) => p.sub)
            .filter((id) => !!id) as string[]
          const subPropertyTerms = await Promise.all(
            subPropertyIds.map((id) => nodes.get(id)),
          )
          return subPropertyTerms
        }),
      )
    )
      .flat()
      .filter((n) => !!n)

    return allProperties as unknown as Node[]
  }

  async getAllTerms() {
    const all = await (await this.nodeStore()).index('by-type').getAll('CLASS')
    return all.filter((term) => {
      return !term.meta?.deprecated
    })
  }
}
