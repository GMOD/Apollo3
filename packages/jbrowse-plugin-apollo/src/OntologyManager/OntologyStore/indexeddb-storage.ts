import { openLocation } from '@jbrowse/core/util/io'
import { IDBPDatabase, IDBPTransaction, openDB } from 'idb/with-async-ittr'

import {
  OntologyDB,
  isOntologyDBEdge,
  isOntologyDBNode,
} from './indexeddb-schema'
import { GraphDocument } from './obo-graph-json-schema'
import OntologyStore from '.'

/** schema version we are currently on, used for the IndexedDB schema open call */
const schemaVersion = 1

export type Database = IDBPDatabase<OntologyDB>

/** open the IndexedDB and create the DB schema if necessary */
export async function openDatabase(dbName: string) {
  // await deleteDB(dbName) // uncomment this to reload every time during development
  return openDB<OntologyDB>(dbName, schemaVersion, {
    upgrade(
      database: IDBPDatabase<OntologyDB>,
      oldVersion: number,
      newVersion: number | null,
      transaction: IDBPTransaction<
        OntologyDB,
        ArrayLike<'nodes' | 'edges'>,
        'versionchange'
      >,
      _event: IDBVersionChangeEvent,
    ): void {
      if (schemaVersion !== 1) {
        throw new Error(
          'now that the schemaVersion is past 1, you need to write some upgrade logic here',
        )
      }
      if (!database.objectStoreNames.contains('meta')) {
        database.createObjectStore('meta')
      }
      if (!database.objectStoreNames.contains('nodes')) {
        database.createObjectStore('nodes', { keyPath: 'id' })
        const nodes = transaction.objectStore('nodes')
        nodes.createIndex('by-label', 'lbl')
        nodes.createIndex('by-type', 'type')
        nodes.createIndex('by-synonym', ['meta', 'synonyms', 'val'])
      }
      if (!database.objectStoreNames.contains('edges')) {
        database.createObjectStore('edges', { autoIncrement: true })
        const edges = transaction.objectStore('edges')
        edges.createIndex('by-subject', 'sub')
        edges.createIndex('by-object', 'obj')
        edges.createIndex('by-predicate', 'pred')
      }
    },
  })
}

/** load a OBO Graph JSON file into a database */
export async function loadOboGraphJson(store: OntologyStore, db: Database) {
  const startTime = Date.now()

  // TODO: using file streaming along with an event-based json parser
  // instead of JSON.parse and .readFile could probably make this faster
  // and less memory intensive
  const oboGraph = JSON.parse(
    await openLocation(store.sourceLocation).readFile('utf8'),
  ) as GraphDocument

  const parseTime = Date.now()

  const [graph, ...additionalGraphs] = oboGraph.graphs ?? []
  if (!graph) {
    return
  }
  if (additionalGraphs.length) {
    throw new Error('multiple graphs not supported')
  }

  try {
    const tx = db.transaction(['meta', 'nodes', 'edges'], 'readwrite')
    await tx.objectStore('meta').clear()
    await tx.objectStore('nodes').clear()
    await tx.objectStore('edges').clear()

    // load nodes
    const nodeStore = tx.objectStore('nodes')
    for (const node of graph.nodes ?? []) {
      if (isOntologyDBNode(node)) {
        await nodeStore.add(node)
      }
    }

    // load edges
    const edgeStore = tx.objectStore('edges')
    for (const edge of graph.edges ?? []) {
      if (isOntologyDBEdge(edge)) {
        await edgeStore.add(edge)
      }
    }

    await tx.done

    // record some metadata about this ontology and load operation
    const tx2 = db.transaction('meta', 'readwrite')
    await tx2.objectStore('meta').add(
      {
        ontologyRecord: {
          name: store.ontologyName,
          version: store.ontologyVersion,
          sourceLocation: store.sourceLocation,
        },
        graphMeta: graph.meta,
        timestamp: String(new Date()),
        schemaVersion,
        timings: {
          overall: Date.now() - startTime,
          load: Date.now() - parseTime,
        },
      },
      'meta',
    )

    await tx2.done
  } catch (e) {
    await db.transaction('meta', 'readwrite').objectStore('meta').clear()
    throw e
  }
  return
}

export async function isDatabaseCompletelyLoaded(db: Database) {
  // since metadata is loaded last, we use it as a signal that all the other data
  // was loaded
  const [meta] = await db.transaction('meta').objectStore('meta').getAll()
  return !!meta
}
