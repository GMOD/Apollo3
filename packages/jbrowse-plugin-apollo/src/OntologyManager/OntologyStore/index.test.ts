import path from 'node:path'

import { beforeAll, describe, expect, it, jest } from '@jest/globals'

import OntologyStore from '.'
import { OntologyClass, isOntologyClass } from '..'

jest.setTimeout(1_000_000_000)

const prefixes = new Map([
  ['SO:', 'http://purl.obolibrary.org/obo/SO_'],
  ['GO:', 'http://purl.obolibrary.org/obo/GO_'],
])

// jsonpath uses an "obj instanceof Object" check in its "query", which fails in
// tests because the mocked indexedDb uses a different scope and thus a
// different "Object". This intercepts calls to "query" in this test and makes
// sure the main scope "Object" is used.
jest.mock('jsonpath', () => {
  const original = jest.requireActual('jsonpath') as typeof import('jsonpath')
  return {
    ...original,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    query: jest.fn((obj: any, pathExpression: string, count?: number) => {
      const newObj =
        obj instanceof Object ? obj : JSON.parse(JSON.stringify(obj))
      return original.query(newObj, pathExpression, count)
    }),
  }
})

let so: OntologyStore

beforeAll(async () => {
  const localPath = path.resolve(__dirname, '../../../test_data/so-v3.1.json')
  so = new OntologyStore(
    'Sequence Ontology',
    'automated testing',
    { locationType: 'LocalPathLocation', localPath },
    { prefixes },
  )
  await so.db
})

describe('OntologyStore', () => {
  it('can load goslim generic', async () => {
    const localPath = path.resolve(
      __dirname,
      '../../../test_data/goslim_generic.json',
    )
    const goslimGeneric = new OntologyStore(
      'Gene Ontology',
      'automated testing',
      { locationType: 'LocalPathLocation', localPath },
      { prefixes },
    )

    expect(await goslimGeneric.termCount()).toMatchSnapshot()

    expect(
      await goslimGeneric.getTermsByFulltext('mitotic nuclear division'),
    ).toMatchSnapshot()
  })

  it('can query SO', async () => {
    expect(await so.termCount()).toMatchSnapshot()
  })
  it('can query SO gene terms and parts of genes', async () => {
    const geneTerms = await so.getTermsWithLabelOrSynonym('gene')
    expect(geneTerms).toMatchSnapshot()
    expect(isOntologyClass(geneTerms[0])).toBe(true)
    const geneParts = await so.getClassesThat(
      'part_of',
      geneTerms as OntologyClass[],
    )
    expect(geneParts).toMatchSnapshot()
  })
  it('can query SO features not part of something else', async () => {
    const topLevelClasses = await so.getClassesWithoutPropertyLabeled(
      'part_of',
      { includeSubProperties: true },
    )
    expect(topLevelClasses.length).toMatchSnapshot()
    expect(topLevelClasses.find((term) => term.lbl === 'mRNA')).toBeUndefined()
    // gene is member_of gene_group, so also doesn't appear here. There doesn't seem to be
    // clarification in SO for when a feature MUST be part_of.
    expect(topLevelClasses.find((term) => term.lbl === 'gene')).toBeUndefined()
  })
  it('can expand subclasses of SO:000039 and still get SO:000039', async () => {
    const expanded = so.expandSubclasses(
      ['http://purl.obolibrary.org/obo/SO_0000039'],
      'is_a',
    )

    const ex = []
    for await (const node of expanded) {
      ex.push(node)
    }
    expect(ex.length).toBeGreaterThan(0)
    expect(ex).toEqual(['http://purl.obolibrary.org/obo/SO_0000039'])
  })
  it('can query valid part_of for match', async () => {
    const parentTypeTerms = await so.getTermsWithLabelOrSynonym('match', {
      includeSubclasses: false,
    })
    // eslint-disable-next-line unicorn/no-array-callback-reference
    const parentTypeClassTerms = parentTypeTerms.filter(isOntologyClass)
    expect(parentTypeClassTerms).toMatchSnapshot()
    const subpartTerms = await so.getClassesThat(
      'part_of',
      parentTypeClassTerms,
    )
    expect(subpartTerms.length).toBeGreaterThan(0)
  })

  it('SO clone_insert_end is among valid subparts of BAC_cloned_genomic_insert', async () => {
    const bcgi = await so.getTermsWithLabelOrSynonym(
      'BAC_cloned_genomic_insert',
      { includeSubclasses: false },
    )
    // eslint-disable-next-line unicorn/no-array-callback-reference
    const bcgiClass = bcgi.filter(isOntologyClass)
    expect(bcgiClass.length).toBe(1)
    expect(bcgiClass[0].lbl).toBe('BAC_cloned_genomic_insert')
    const subpartTerms = await so.getClassesThat('part_of', bcgiClass)
    expect(subpartTerms.length).toBeGreaterThan(0)
    expect(subpartTerms.find((t) => t.lbl === 'clone_insert_end')).toBeTruthy()
    expect(subpartTerms).toMatchSnapshot()
  })
})
