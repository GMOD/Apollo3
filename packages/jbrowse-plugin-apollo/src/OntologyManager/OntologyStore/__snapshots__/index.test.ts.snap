// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OntologyStore can load SO 1`] = `2629`;

exports[`OntologyStore can load goslim aspergillus 1`] = `187`;

exports[`OntologyStore can load goslim generic 1`] = `248`;
